sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -47,18 +47,20 @@ class OrcDeserializer(
       }.toArray
   }
 
-  private val validColIds = requestedColIds.filterNot(_ == -1)
-
   def deserialize(orcStruct: OrcStruct): InternalRow = {
-    var i = 0
-    while (i < validColIds.length) {
-      val value = orcStruct.getFieldValue(validColIds(i))
-      if (value == null) {
-        resultRow.setNullAt(i)
-      } else {
-        fieldWriters(i)(value)
+    var fieldWriterIndex = 0
+    var targetColumnIndex = 0
+    while (targetColumnIndex < requestedColIds.length) {
```

Comment:
now this loop looks a little complicated, as there are 2 indexes.

How about we make the `fieldWriters` has the same size as `requiredSchema`? i.e.
```
private val fieldWriters: Array[WritableComparable[_] => Unit] = requiredSchema.zipWithIndex.map {
  case (f, index) =>
    if (requestedColIds(index) == -1) {
      null
    } else {
      // create writer
    }
}
```

Then this loop becomes
```
var i = 0
while (i < fieldWriters.length) {
  ...
}
```

Reply:
I considered this as well. The currently implemented version allowed me to keep the change more isolated (a single method only), which seemed less risky and easier to review.

However, I agree that changing `fieldWriters` as well is probably better for readability in the long run. I’m definitely open to implementing it this way if there’s consensus that this is better.

(Minor note: if we do switch to that implementation, I would make the missing field writers None instead of null.)

Reply:
In performance critical code path, we should prefer the java style, i.e. using null instead of `Option` to avoid overhead.